mne/epochs.py

```diff
@@ -2640,6 +2701,121 @@ def _get_epoch_from_raw(self, idx, verbose=None):
                                             self.reject_by_annotation)
         return data
 
+    @property
+    def annotations(self):  # noqa: D102
+        return self._annotations
+
+    @verbose
+    def _set_annotations(self, annotations, on_missing='raise'):
+        """Setter for Epoch annotations from Raw.
+
+        This private function simply copies over Raw annotations, and
+        does not handle offsetting the times based on first_samp
+        or measurement dates, since that is expected to occur in
+        Raw.set_annotations().
+
+        Parameters
+        ----------
+        annotations : instance of mne.Annotations | None
+            Annotations to set. If None, the annotations is defined
+            but empty.
+        %(on_missing_ch_names)s
+
+        Returns
+        -------
+        self : instance of Raw
+            The raw object with annotations.
+        """
+        meas_date = _handle_meas_date(self.info['meas_date'])
+        if annotations is None:
+            self._annotations = None
+        else:
+            _validate_type(annotations, Annotations, 'annotations')
+            new_annotations = annotations.copy()
+            new_annotations._prune_ch_names(self.info, on_missing)
+            self._annotations = new_annotations
+
+        return self
+
+    def get_epoch_annotations(self):
+        """Return a list of raw annotations per epoch.
+
+        Returns
+        -------
+        epoch_annots : list
+            A list (with length equal to number of epochs) of annotations per
+            epoch. Each annotation is stored as (onset, duration, description),
+            where the onset is now relative to the epoch, rather then in
+            continuous time.
+        """
+        events = self.events
+
+        # create a list of annotations for each epoch
+        epoch_annot_list = [[]] * len(events)
```

Comment:
This actually doesn't work because it creates a pointer in RAM to the same list, rather then a separate list per epoch.

`epoch_annot_list = [[] for _ in range(len(events))]` is required.

Reply:
Oh interesting! Could do `[list()] * len(events)` then?

Reply:
Nope unfortunately that actually also doesn't work. Some weird Python magic. I think, if you do anything like:

```
[...] * len(N)
```

it allocates only 1 list and then points to that. I think maybe if you set it to be `None` inside, it bypasses this issue, but then Imo we might as well be explicit w/ the soln I found on SO.

Reply:
Indeed `[[]] * len(events)` will make a list of length N with identical (in memory) lists:
```
>>> x = [[]] * 3
>>> x[0].append(1)
>>> x
[[1], [1], [1]]
```